libs/providers/flagd/src/lib/service/grpc/service.ts

```diff
@@ -2,111 +2,238 @@ import * as grpc from '@grpc/grpc-js';
+    if (message.data) {
+      const parsedMessage = Struct.toJson(message.data) as FlagChangeMessage;
+      this.logger?.debug(`${FlagdProvider.name}: got message: ${JSON.stringify(parsedMessage, undefined, 2)}`);
+      if (parsedMessage.flags) {
+          // remove each changed key from cache
+          Object.keys(parsedMessage.flags).forEach((key) => {
+            if (this._cache?.delete(key)) {
+              this.logger?.info(`${FlagdProvider.name}: evicted key: ${key} from cache.`);
+            }
+          });
+      }
+    }
+  }
+
+  private handleError(reject: (reason?: any) => void) {
+    this.logger?.error(`${FlagdProvider.name}: streaming connection error, will attempt reconnect...`);
+    this._cache?.clear();
+    this._streamAlive = false;
+
+    // if we haven't reached max attempt, reconnect after backoff
+    if (this._streamConnectAttempt <= this._maxEventStreamRetries) {
+      this._streamConnectAttempt++;
+      setTimeout(() => {
+        this._streamConnectBackoff = this._streamConnectBackoff * 2;
+        this.connectStream();
+      }, this._streamConnectBackoff);
+    } else {
+
+      // after max attempts, give up
+      const errorMessage = `${FlagdProvider.name}: max stream connect attempts (${this._maxEventStreamRetries} reached)`;
+      this.logger?.error(errorMessage);
+      reject(new Error(errorMessage));
+    }
+  }
+
+  private handleComplete() {
+    this.logger?.info(`${FlagdProvider.name}: streaming connection closed gracefully`);
+    this._cache?.clear();
+    this._streamAlive = false;
+  }
+
+  private objectParser = (struct: Struct) => {
+    if (struct !== undefined) {
+      return Struct.toJson(struct);
     } else {
       throw new ParseError('Object value undefined or missing.');
     }
+  };
+
+  private async resolve<T extends FlagValue, Rq extends AnyRequest, Rs extends AnyResponse>(
+    resolver: (request: AnyRequest) => UnaryCall<Rq, Rs>,
+    flagKey: string,
+    context: EvaluationContext,
+    logger: Logger,
+    parser?: (struct: Struct) => PbJsonValue
+  ): Promise<ResolutionDetails<T>> {
+    if (this._cacheActive) {
+      const cached = this._cache?.get(flagKey);
+      if (cached) {
+        return cached as ResolutionDetails<T>;
+      }
+    }
+
+    // invoke the passed resolver method
+    const { response } = await resolver
+      .call(this._client, { flagKey, context: this.convertContext(context, logger) })
+      .then((resolved) => resolved, this.onRejected);
+
+    const resolved = {
+      // invoke the parser method if passed
+      value: parser ? parser.call(this, response.value as Struct) : response.value,
+      reason: response.reason,
+      variant: response.variant,
+    } as ResolutionDetails<T>;
+
+    if (this._cacheActive && response.reason === StandardResolutionReasons.STATIC) {
+      // cache this static value
+      this._cache?.set(flagKey, { ...resolved, reason: StandardResolutionReasons.CACHED });
```

Comment:
This differs from the other implementations in that it sets the reason as CACHED when storing rather than setting it on the retrieved value from cache. This may be fine but thought it was worth pointing out

Reply:
Interesting thought. I will think about this a bit.

Reply:
It may be slightly better to set `CACHED` after the value is retrieved. However, I can't think of a scenario where the behavior would be different between the two different implementations.

Reply:
updated for consistency.